edb/pgsql/resolver/context.py

```diff
@@ -0,0 +1,174 @@
+#
+# This source file is part of the EdgeDB open source project.
+#
+# Copyright 2008-present MagicStack Inc. and the EdgeDB authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from __future__ import annotations
+
+from copy import deepcopy
+from typing import *
+import enum
+
+from edb.common import compiler
+from edb.schema import schema as s_schema
+
+
+class Scope:
+    """
+    Information about that objects are visible at a specific point in an
+    SQL query.
+
+    Scope is modified during resolving of a query, when new tables are
+    discovered in FROM or JOIN or new columns declared in SELECT's projection.
+
+    After a query is done resolving, resulting relations are extracted from its
+    scope and inserted into parent scope.
+    """
+
+    rel: Table
+    "Current relation where ResTargets are added to"
+
+    join_relations: List[Table]
+    "Current relation where Join relations are added to"
+
+    tables: List[Table]
+    """Tables visible in this scope"""
+
+    ctes: List[CTE]
```

Comment:
Yeah, they could but then I would we duplicating stored table name in `Table.name` and as key in `tables`. And tables would have to be stored under two keys, because table can be referred with its table name and also with an alias.

Having it as a list is not really slower in most cases, because number of tables will very rarely exceed 3.